control-plane-operator/hostedclusterconfigoperator/controllers/drainer/drainer.go

```diff
@@ -0,0 +1,162 @@
+	guestClusterKubeClient kubeclient.Interface
+	releaseProvider        releaseinfo.Provider
+	upsert.CreateOrUpdateProvider
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling")
+
+	node := &corev1.Node{}
+	err := r.guestClusterClient.Get(ctx, req.NamespacedName, node)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error getting node")
+		return ctrl.Result{}, err
+	}
+
+	if node.Annotations[DesiredDrainerAnnotationKey] != node.Annotations[LastAppliedDrainerAnnotationKey] {
+		if err = r.handleNodeDrainRequest(ctx, node, node.Annotations[DesiredDrainerAnnotationKey]); err != nil {
+			// TODO (jerzhang): properly handle failure and error detection
+			// Right now for individual failures we will endlessly retry
+			// Instead we should have some timeout (based on NodeDrainTimeout?)
+			// and error reporting to the nodepool status
+			log.Error(err, "failed to drain node")
+			return ctrl.Result{RequeueAfter: drainFailureRequeueDuration}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) handleNodeDrainRequest(ctx context.Context, node *corev1.Node, desiredState string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// TODO (jerzhang): this name is deterministic, but should make this a const somewhere
+	daemonPodOnNodeName := fmt.Sprintf("machine-config-daemon-%s", node.Name)
+	drainer := &drain.Helper{
+		Client:                          r.guestClusterKubeClient,
+		Force:                           true,
+		IgnoreAllDaemonSets:             true,
+		DeleteEmptyDirData:              true,
+		GracePeriodSeconds:              -1,
+		SkipWaitForDeleteTimeoutSeconds: 300,
+		Timeout:                         90 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			log.Info("Pod drained", "Action", verbStr, "Namespace", pod.Namespace, "Pod name", pod.Name)
+		},
+		AdditionalFilters: []drain.PodFilter{
+			func(pod corev1.Pod) drain.PodDeleteStatus {
+				if pod.Name == daemonPodOnNodeName {
+					return drain.MakePodDeleteStatusSkip()
+				}
+				return drain.MakePodDeleteStatusOkay()
+			},
+		},
+		// TODO (jerzhang): properly handle logging here, although this seems to work
+		Out:    writer{log.Info},
+		ErrOut: writer{log.Info},
+		Ctx:    ctx,
+	}
+
+	desiredVerb := strings.Split(desiredState, "-")[0]
+	switch desiredVerb {
+	case DrainerStateUncordon:
+		log.Info("Performing uncordon on node", "name", node.Name)
+		if err := drain.RunCordonOrUncordon(drainer, node, false); err != nil {
+			return fmt.Errorf("uncordon failed: %w", err)
+		}
+	case DrainerStateDrain:
+		log.Info("Performing cordon on node", "name", node.Name)
+		if err := drain.RunCordonOrUncordon(drainer, node, true); err != nil {
```

Comment:
this is redundantly calling RunCordonOrUncordon and within applyDrain again.

Reply:
ack, will remove one or the other